elk/utils_evaluation/ccs.py

```diff
@@ -217,3 +203,74 @@ def fit(self, data: list, label):
     def score(self, data: list, label, getloss=False):
         self.validate_data(data)
         return self.get_accuracy(self.best_theta, data, label, getloss)
+
+    def train_loop_full_batch(self, x0, x1, theta):
+        """
+        Performs a full batch training loop. Modifies theta in place.
+        """
+        optimizer = torch.optim.AdamW(
+            [theta], lr=self.learning_rate, weight_decay=self.weight_decay
+        )
+
+        # Start training (full batch)
+        for _ in range(self.num_epochs):
+
+            # project onto theta
+            z0, z1 = x0.mm(theta.T), x1.mm(theta.T)
+
+            # sigmoid to get probability
+            p0, p1 = torch.sigmoid(z0), torch.sigmoid(z1)
+
+            # get the corresponding loss
+            loss = self.get_loss(p0, p1)
+
+            # update the parameters
+            optimizer.zero_grad()
+            loss.backward()
+            optimizer.step()
+
+            # with torch.no_grad():
+            #     theta /= torch.norm(theta)
```

Comment:
This should probably just be deleted if we're not using it?

Reply:
Done